### PR TITLE
Support Return Composite Type for Func Call

### PR DESCRIPTION
**Pull request by Quasimewdo**
- Able to create composite type
- Able to call on composite type
- Able to return composite type

### CODE REVIEW

File: tests/test_type.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        type.create_type(Person, "Person", db)
+    assert 'type "person" already exists\n' in str(exc_info.value)
+
+
+def test_type_drop(db: gp.Database):
+    class Person:
+        _first_name: str
+        _last_name: str
+
+    type.create_type(Person, "Person", db)
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply tests/test_type.py:30]: I am also thinking about it, maybe let the user to create the return type if it doesn't exist.

File: tests/test_func.py [hunk 2]
```diff
@@ -201,3 +201,24 @@ def my_sum(val_list: List[int]) -> int:
...
+        ) -> None:
+            self._first_name = first_name
+            self._last_name = last_name
+
+    @gp.create_function
+    def create_person(first: str, last: str) -> Person:
+        return {"first_name": first, "last_name": last}
+
+    for row in create_person("'Amy'", "'An'", as_name="result", db=db).to_table().fetch():
+        print(row["result"])
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply tests/test_func.py]: Fixed

File: tests/test_func.py [hunk 3]
```diff
@@ -201,3 +204,52 @@ def my_sum(val_list: List[int]) -> int:
...
+
+    @gp.create_function
+    def create_person(first: str, last: str) -> Person:
+        return {"first_name": first, "last_name": last}
+
+    for row in create_person("'Amy'", "'An'", db=db).to_table().fetch():
+        assert row["first_name"] == "Amy" and row["last_name"] == "An"
+
+
+def test_func_comp_type_column(db: gp.Database):
```
Comment (beeender) [inline tests/test_func.py:222]: This test is not quite different from `test_func_return_comp_type`, what is the main purpose of this test? should it have a different name?
Is the original purpose of this test to test a composite-type column? but the column type is `int` I think?

File: tests/test_type.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    class Person:
+        _first_name: str
+        _last_name: str
+
+    class Couple:
+        _first_person: Person
+        _second_person: Person
+
+    def create_couple() -> Couple:
+        return Couple
```
Comment (beeender) [inline tests/test_type.py:74]: it should return a `Couple` object instead of a `Couple` type.
Change it to `Couple()`

File: tests/test_type.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    class Person:
+        _first_name: str
+        _last_name: str
+
+    type.create_type(Person, db, as_name="Person", is_temp=False)
+    type.drop_type("Person", db)
+    type.create_type(Person, db, as_name="Person", is_temp=False)
+
+
+def test_create_type_rec(db: gp.Database):
```
Comment (beeender) [inline tests/test_type.py:64]: does `rec` mean `recursive` here?

File: greenplumpython/table.py [hunk 6]
```diff
@@ -1,6 +1,8 @@
 from typing import TYPE_CHECKING, Iterable, List, Optional, Tuple
 from uuid import uuid4
 
+from tabulate import tabulate
```
Comment (beeender) [inline greenplumpython/table.py:4]: move to a separate PR.

File: tests/test_type.py (hunk 5 above)
Comment (Quasimewdo) [reply tests/test_type.py:64]: yes
```suggestion
def test_create_type_recursive(db: gp.Database):
```

**Comment by Quasimewdo** [comment]
> There should be a type `Couple` in the signature, while an object `Couple()` in the return statement.

`    def create_couple() -> Couple:
        return Couple()`

File: greenplumpython/type.py [hunk 7]
```diff
@@ -10,15 +15,50 @@
...
 
 
+# TODO : Add tests for all function
+def create_type(class_type, db, as_name=None, is_temp=True) -> str:
+    type_name = "type_" + uuid4().hex if as_name is None else as_name
+    temp_str = "pg_temp." if is_temp else ""
+    att_type_str = ",\n\t\t\t\t".join(
+        [
+            f"""
+            {re.sub(r"^_+|_+$", "", name)} {to_pg_type(type_t, db)}
```
[the remark this answers is not in the document]
Comment (Quasimewdo) [reply greenplumpython/type.py:25]: There was an error when having underscore, but now it is ok
```suggestion
            {name} {to_pg_type(type_t, db)}
```

File: tests/test_func.py [hunk 8]
```diff
@@ -51,16 +51,18 @@ def my_max(a: int, b: int) -> int:
...
-	def my_min(a: int, b: int) -> int:
-		if a < b:
-			return a
-		else:
-			return b
-
-	for row in my_min(1, 2, as_name="result", db=db).to_table().fetch():
-		assert row["result"] == min(1, 2)
-		assert row["result"] == inspect.unwrap(my_min)(1, 2)
+    @gp.create_function
```
[the remark this answers is not in the document]
Comment (beeender) [reply tests/test_func.py]: not recovered :(

**Review by beeender — COMMENTED** [review]
Please fix those minor issues then we are good to go !
